pandas/_libs/tslibs/conversion.pyx

```diff
@@ -554,15 +615,16 @@ cdef inline void localize_tso(_TSObject obj, tzinfo tz):
             # static/fixed tzinfo; in this case we know len(deltas) == 1
             # This can come back with `typ` of either "fixed" or None
             dt64_to_dtstruct(obj.value + deltas[0], &obj.dts)
-        elif typ == 'pytz':
-            # i.e. treat_tz_as_pytz(tz)
-            pos = trans.searchsorted(obj.value, side='right') - 1
-            tz = tz._tzinfos[tz._transition_info[pos]]
-            dt64_to_dtstruct(obj.value + deltas[pos], &obj.dts)
-        elif typ == 'dateutil':
-            # i.e. treat_tz_as_dateutil(tz)
+        elif typ == 'pytz' or typ == 'dateutil':
             pos = trans.searchsorted(obj.value, side='right') - 1
+            if typ == 'pytz':
+                tz = tz._tzinfos[tz._transition_info[pos]]
+            # Adjust value if fold was supplied
+            if fold == 1:
+                pos = _adjust_tsobject_for_fold(obj, trans, deltas, pos, fold)
             dt64_to_dtstruct(obj.value + deltas[pos], &obj.dts)
+
+            obj.fold = _infer_tsobject_fold(obj, trans, deltas, pos)
```

Comment:
@mroeschke 
I propose we move the value adjustment and fold inferring functionality to two functions. Moving the larger `if` block is problematic, because the details differ between `localize_tso`, `convert_datetime_to_tsobject`, `create_tsobject_tz_using_offset`. This way we get rid of duplicate logic without needlessly calling the functions or coupling them with the calling functions too tightly.

Reply:
Sounds good. Thanks for simplifying it to this point.